flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/CatalogITCase.java

```diff
@@ -61,9 +70,54 @@ public void testDropCatalog() {
 		assertFalse(tableEnv.getCatalog(name).isPresent());
 	}
 
+	@Test
+	public void testCreateCatalogFromUserClassLoader() throws Exception {
+		final String className = "UserCatalogFactory";
+		URLClassLoader classLoader = ClassLoaderUtils.withRoot(temporaryFolder.newFolder())
+			.addResource("META-INF/services/org.apache.flink.table.factories.TableFactory", "UserCatalogFactory")
+			.addClass(
+				className,
+				"import org.apache.flink.table.catalog.GenericInMemoryCatalog;\n" +
+					"import org.apache.flink.table.factories.CatalogFactory;\n" +
+					"import java.util.Collections;\n" +
+					"import org.apache.flink.table.catalog.Catalog;\n" +
+					"import java.util.HashMap;\n" +
+					"import java.util.List;\n" +
+					"import java.util.Map;\n" +
+					"\tpublic class UserCatalogFactory implements CatalogFactory {\n" +
+					"\t\t@Override\n" +
+					"\t\tpublic Catalog createCatalog(\n" +
+					"\t\t\t\tString name,\n" +
+					"\t\t\t\tMap<String, String> properties) {\n" +
+					"\t\t\treturn new GenericInMemoryCatalog(name);\n" +
+					"\t\t}\n" +
+					"\n" +
+					"\t\t@Override\n" +
+					"\t\tpublic Map<String, String> requiredContext() {\n" +
+					"\t\t\tHashMap<String, String> hashMap = new HashMap<>();\n" +
+					"\t\t\thashMap.put(\"type\", \"userCatalog\");\n" +
+					"\t\t\treturn hashMap;\n" +
+					"\t\t}\n" +
+					"\n" +
+					"\t\t@Override\n" +
+					"\t\tpublic List<String> supportedProperties() {\n" +
+					"\t\t\treturn Collections.emptyList();\n" +
+					"\t\t}\n" +
+					"\t}"
+			).build();
+
+		try (TemporaryClassLoaderContext context = TemporaryClassLoaderContext.of(classLoader)) {
```

Comment:
This case can run successfully even without the commit `[FLINK-18419] Create catalog in TableEnvironment using user ClassLoader`. Because `TemporaryClassLoaderContext.of` method sets the user classloader to the current thread, and `TableFactoryService#discoverFactories` method will use the classloader in the current thread to find factories.

Reply:
I don't think so. The problem is the line in `SqlToOperationConverter:557`.
```
		final CatalogFactory factory =
			TableFactoryService.find(CatalogFactory.class, properties, this.getClass().getClassLoader());
```
 It is using the class classloader, not the context classloader. I could've just changed that to use the context classloader, but I found it wrong that the catalog is instantiated in `SqlToOperationConverter` and it would not use the classloader provided in `EnvironmentSettings` in the future.

Reply:
I can reproduce the exception after revert the fixing changes.

Reply:
you are right.`SqlToOperationConverter:557` is the root cause. I find the test case does not covert the full change path (not the whole commit), such as I remove the line in `TableEnvironmentImpl#1052`,
```
final CatalogFactory factory = TableFactoryService.find(
				CatalogFactory.class,
				properties);
```
The test case can also run successfully. Do we need build `StreamExecutionEnvironmentImpl` with user classloader directly to verify the change.

Reply:
I see your point. Nevertheless your approach would not test a different code path. That the context classloader is used. This behaviour is important from the sql-client perspective.

I lean slightly towards the current approach, as it uses only the user facing APIs. It tests what is expected in sql-client. Imo, a test that the classloader from EnvironmentSettings is used should be part of https://issues.apache.org/jira/browse/FLINK-15635 

WDYT?

Reply:
ok, let FLINK-15635 fully support this feature